core/src/util_regex.cpp

```diff
@@ -0,0 +1,352 @@
+  assert(matchLen != -1); // error
+  if (matchLen == 0) {
+    return 0; // Normal case return: no match
+  }
+  std::u32string rustr; // replacement
+  if (fromList.empty()) {
+    // Normal case: not a map.
+    // This replace will apply $1, $2 etc.
+    rustr = to;
+  } else {
+    // we actually need the group(1) string here.
+    // this is only the content in parenthesis ()
+    char *group1 = RegexGroup1(patstr.c_str(), instr.c_str());
+    assert(group1 != nullptr);
+    const std::string group1str(group1);
+    const std::u32string match32 = convert<char, char32_t>(group1str);
+    free(group1);
+    // Now we're ready to do the actual mapping.
+
+    // 1., we need to find the index in the source set.
+    auto matchIndex = findIndex(match32, fromList);
+    assert(matchIndex != -1L); // This indicates that the regex and the fromList are out of sync.
+    // we already asserted on load that the from and to sets have the same cardinality.
+
+    // 2. get the target string
+    // we use the same matchIndex that was just found
+    // 3. update the UnicodeString for replacement
+    rustr = toList.at(matchIndex);
+  }
+  std::string rstr = convert<char32_t,char>(rustr);
+  // here we replace the match output.
+  char *out = RegexSubstitute(patstr.c_str(), instr.c_str(), rstr.c_str());
+  assert(out != nullptr);
+  std::string outstr(out);
+  free(out);
+  output = convert<char, char32_t>(outstr);
+  // output includes all of 'input', but modified. Need to substring it.
+  /** code units */
+  const auto matchStart = input.length() - matchLen;
+  // remove the unmatched prefix.
+  output.erase(0, matchStart);
+
+  return matchLen;
+#else
+  assert(fPattern);
+  // TODO-LDML: This entire section may have too many conversions and copies. Could be optimized.
+  UErrorCode status = U_ZERO_ERROR;
+  const std::u16string matchstr = km::core::kmx::u32string_to_u16string(input);
+  icu::UnicodeString matchustr  = icu::UnicodeString(matchstr.data(), (int32_t)matchstr.length());
+  // TODO-LDML: create a new Matcher every time. These could be cached and reset.
+  std::unique_ptr<icu::RegexMatcher> matcher(fPattern->matcher(matchustr, status));
+  if (!UASSERT_SUCCESS(status)) {
+    return 0;
+  }
+
+  if (!matcher->find(status)) { // i.e. matches somewhere, in this case at end of str
+    return 0; // Normal case return: no match
+  }
+
+  // Note: this is UTF-16 len, not UTF-32 len.
+  int32_t matchStart = matcher->start(status);
+  int32_t matchEnd   = matcher->end(status);
+  if (!UASSERT_SUCCESS(status)) {
+    return 0; // TODO-LDML: return error
+  }
+  // extract..
+  const icu::UnicodeString substr = matchustr.tempSubStringBetween(matchStart, matchEnd);
+  // preflight to UTF-32 to get length
+  UErrorCode substrStatus = U_ZERO_ERROR; // throwaway status
+  // we need the UTF-32 matchLen for our return.
+  auto matchLen = substr.toUTF32(nullptr, 0, substrStatus);
+
+  // should have matched something.
+  assert(matchLen > 0);
```

Comment:
In the `KM_NO_ICU` section, we `assert(matchLen != -1);`.  Why the difference here in treatment of `0`?

Reply:
- because matchLen in the other section is used to communicate an error. In this code branch, we check for errors at these lines and others.  

https://github.com/keymanapp/keyman/blob/ca34550a8d3c3e88ede1b579d7ad678c4cfcf181/core/src/util_regex.cpp

https://github.com/keymanapp/keyman/blob/ca34550a8d3c3e88ede1b579d7ad678c4cfcf181/core/src/util_regex.cpp